[torchgen] Support multiple namespace in NativeFunctions.h

larryliu0820 opened the pull request:
Summary:
This is a follow up to #78015. This PR
* introduces namespace logic for generating `NativeFunctions.h`.
* adds helper function to extract namespace from string
* relaxes the constraint on the levels we support for custom kernel namespace to 2

Test Plan:
Yaml entry:
```
- func: unsqueeze.out(Tensor(a) self, int dim, *, Tensor(a!) out) -> Tensor(a!)
  variants: function
  device_check: NoCheck
  dispatch:
    CPU: custom_1::custom_2::unsqueeze
```

Generated `NativeFunctions.h`:

```
namespace custom_1 {
namespace custom_2 {
namespace native {
    TORCH_API at::Tensor & unsqueeze(const at::Tensor & self, int64_t dim, at::Tensor & out);
} // namespace native
} // namespace custom_2
} // namespace custom_1

```

Differential Revision: D37198111

dbort commented on torchgen/utils.py:
```diff
@@ -396,3 +396,52 @@ def _format(
+        return self.entity_name_
+
+    # Only allow certain level of namespaces
+    def get_namespaces(self, max_level: int = 1, default: str = "") -> str:
```
Please add a docstring describing how/when the default is used. Also document the format of the return string (i.e., a `::`-separated C++ namespace string, without a leading `::`)

In fact since this does return a single C++ namespace string, consider calling it the singular `get_namespace` or even `get_cpp_namespace` to make it more clear that it's C++ specific and doesn't return a list.

dbort commented on torchgen/utils.py:
```diff
@@ -396,3 +396,52 @@ def _format(
     return f"{start}{indent_str}{body}{end}"
+
+
+class NamespaceHelper:
```
This class might be a good place to check that the namespace provided in the YAML file is a valid C++ namespace; i.e., elements are in `/[a-zA-Z_][a-zA-Z_0-9]*/`. That would help trigger error early, when operating on the YAML files, instead of later on when building the generated code files.

larryliu0820 commented on torchgen/utils.py:
```diff
@@ -396,3 +396,52 @@ def _format(
+    def __init__(self, namespace_str: str):
+        # cpp_namespace can be a colon joined string such as torch::lazy
+        cpp_namespaces = namespace_str.split("::")
+        self.prologue_ = "\n".join([f"namespace {n} {{" for n in cpp_namespaces])
```
Yeah good point. I was making changes on an existing `NamespaceHelper` and apparently the old class has a different assumption on the input. Let me refactor this and add some unit tests.

iseeyuan commented:
> relaxes the constraint on the levels we support for custom kernel namespace to 2

Nice. Maybe the example in the summary could reflect this improvement.

bdhirsh commented on torchgen/gen.py:
```diff
@@ -1374,6 +1345,81 @@ def flatten_pre_group(
+            namespace = (
+                backend_metadata.cpp_namespace
+                if backend_metadata
+                else DEFAULT_KERNEL_NAMESPACE
```
So, reading this, it looks like we're implying full flexibility that for a given operator, it might have a different namespace for different backends.

I don't think that's actually true in practice, right? We don't even have a way to express that in the yaml syntax - from the yaml syntax, you'd expect that the cpp_namespace is a property of the `NativeFunction`, and not the backend.

The per-backend part comes mostly users of `gen_backend_stubs.py` and `gen_lazy_tensor.py`; if your a backend that uses one of those files to run codegen, you can specify a different namespace for your operator than the "default" specified from `native_functions.yaml` (so in one of your recent PR's, we slotted it into the `BackendIndex`). 

I wouldn't expect multiple namespaces to show up for a single operator for anything that ever invokes `gen.py` though (do you agree that's the case for mobile too?)

If that's the case, then maybe just assert that find all valid `BackendMetadata` objects for a given `NativeFunction`, and assert that their namespaces are all the same.

larryliu0820 replied:
Right now we don't have such use case, but wondering if we can do:

```
- func: cat.out(Tensor[] tensors, int dim=0, *, Tensor(a!) out) -> Tensor(a!)
  dispatch:
    CPU: cpu::cat_out
    CUDA: cuda::cat_out
```

bdhirsh replied:
it feels a bit like unnecessary complexity to me haha (I don't think the codegen has logic to parse a namespace out of the kernel name today). But if that's important, we could probably do it.

bdhirsh commented on torchgen/gen.py:
```diff
@@ -1450,27 +1496,15 @@ def gen_aggregated_headers(
-            "NativeFunctions_declarations": list(
-                concatMap(
-                    # Convert to a set first to remove duplicate kernel names.
-                    # Backends are allowed to repeat kernel names; only generate the declaration once!
```
It looks like you lost this comment when moving the logic over - do you mind moving it into `get_native_function_declarations`?

bdhirsh commented on torchgen/gen.py:
```diff
@@ -1374,6 +1345,81 @@ def flatten_pre_group(
+                dest.compute_native_function_declaration(f, backend_idx)
+            )
+
+    for namespace, kernels in ns_grouped_kernels.items():
```
Actually, I think this might create duplicate declarations (see https://github.com/pytorch/pytorch/pull/79733/files#r915842429). `kernels` is just a list of entries (one generated for every valid `(NativeFunction, BackendIndex)` pair), so it might have more than one entry for a given `NativeFunction`.

The comment basically suggests fixing that by taking your list (`kernels`), converting to a set to remove duplicates (using `OrderedDict.from_keys` to the codegen output is deterministic), and converting back to a list

larryliu0820 commented on torchgen/gen.py:
```diff
@@ -1374,6 +1345,81 @@ def flatten_pre_group(
+        ns_helper = NamespaceHelper(
+            namespace_str=namespace,
+            entity_name="",
+            max_level=3,
```
`torch::executor::native` is the convention for native functions now

bdhirsh reviewed (approved):
lgtm!

larryliu0820 commented:
@pytorchbot merge

larryliu0820 committed:
[torchgen] Support multiple namespace in NativeFunctions.h (#79733)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/79733

This is a follow up to #78015. This PR
* introduces namespace logic for generating `NativeFunctions.h`.
* adds helper function to extract namespace from string
* relaxes the constraint on the levels we support for custom kernel namespace to 2

Test Plan:
Yaml entry:
```
- func: unsqueeze.out(Tensor(a) self, int dim, *, Tensor(a!) out) -> Tensor(a!)
  variants: function
  device_check: NoCheck
  dispatch:
    CPU: custom::unsqueeze
```

Generated `NativeFunctions.h`:

```
namespace custom {
namespace native {
    TORCH_API at::Tensor & unsqueeze(const at::Tensor & self, int64_t dim, at::Tensor & out);
} // namespace native
} // namespace custom

```

Differential Revision: D37198111

fbshipit-source-id: 6f311290c26011ebd7a1ad592e4bfd44409802ef

larryliu0820 commented:
@pytorchbot merge

larryliu0820 commented:
@pytorchbot merge
